File: flutter_app/lib/ear/ear_seed_models.dart
/// 练耳题库中一道四选一 MCQ（模式 A / B）。
class EarMcqOption {
  const EarMcqOption({required this.key, required this.label});

  final String key;
  final String label;
}

/// 从 `ear_seed_v1.json` 解析后的题目。
class EarBankItem {
  const EarBankItem({
    required this.id,
    required this.mode,
    required this.questionType,
    required this.promptZh,
    required this.options,
    required this.correctOptionKey,
    this.root,
    this.targetQuality,
    this.musicKey,
    this.progressionRoman,
    this.hintZh,
  });

  final String id;
  final String mode;
  final String questionType;
  final String promptZh;
  final List<EarMcqOption> options;
  final String correctOptionKey;

  /// 模式 A：`single_chord_quality`
  final String? root;
  final String? targetQuality;

  /// 模式 B：`progression_recognition`
  final String? musicKey;
  final String? progressionRoman;

  final String? hintZh;

  static EarBankItem? tryParse(Object? raw) {
    if (raw is! Map) return null;
    final id = raw['id'];
    final mode = raw['mode'];
    final qt = raw['question_type'];
    final prompt = raw['prompt_zh'];
    final correct = raw['correct_option_key'];
    if (id is! String ||
        mode is! String ||
        qt is! String ||
        prompt is! String ||
        correct is! String) {
      return null;
    }
    final optRaw = raw['options'];
    if (optRaw is! List) return null;
    final options = <EarMcqOption>[];
    for (final o in optRaw) {
      if (o is! Map) continue;
      final k = o['key'];
      final lb = o['label'];
      if (k is! String || lb is! String) continue;
      options.add(EarMcqOption(key: k, label: lb));
    }
    if (options.length < 2) return null;

    final root = raw['root'];
    final tq = raw['target_quality'];
    final mk = raw['music_key'];
    final pr = raw['progression_roman'];
    final hint = raw['hint_zh'];

    return EarBankItem(
      id: id,
      mode: mode,
      questionType: qt,
      promptZh: prompt,
      options: options,
      correctOptionKey: correct,
      root: root is String ? root : null,
      targetQuality: tq is String ? tq : null,
      musicKey: mk is String ? mk : null,
      progressionRoman: pr is String ? pr : null,
      hintZh: hint is String ? hint : null,
    );
  }
}

/// 根 JSON 文档。
class EarSeedDocument {
  const EarSeedDocument({required this.bankA, required this.bankB});

  final List<EarBankItem> bankA;
  final List<EarBankItem> bankB;

  static EarSeedDocument? tryParseMap(Map<String, dynamic> map) {
    final banks = map['banks'];
    if (banks is! Map) return null;
    final aRaw = banks['A'];
    final bRaw = banks['B'];
    final a = <EarBankItem>[];
    final b = <EarBankItem>[];
    if (aRaw is List) {
      for (final x in aRaw) {
        final item = EarBankItem.tryParse(x);
        if (item != null) a.add(item);
      }
    }
    if (bRaw is List) {
      for (final x in bRaw) {
        final item = EarBankItem.tryParse(x);
        if (item != null) b.add(item);
      }
    }
    if (a.isEmpty && b.isEmpty) return null;
    return EarSeedDocument(bankA: a, bankB: b);
  }
}

File: swift_app/Sources/Features/Ear/EarSeedModels.swift
import Foundation

public struct EarMcqOption: Decodable, Hashable, Sendable {
    public let key: String
    public let label: String

    public init(key: String, label: String) {
        self.key = key
        self.label = label
    }
}

public struct EarBankItem: Decodable, Hashable, Sendable, Identifiable {
    public let id: String
    public let mode: String
    public let questionType: String
    public let promptZh: String
    public let options: [EarMcqOption]
    public let correctOptionKey: String
    public let root: String?
    public let targetQuality: String?
    public let musicKey: String?
    public let progressionRoman: String?
    public let hintZh: String?
    /// 程序化和弦听辨：与试听一致的 6→1 弦品格（闷弦为 `-1`）；JSON 题库可无此字段。
    public let playbackFretsSixToOne: [Int]?

    enum CodingKeys: String, CodingKey {
        case id
        case mode
        case questionType = "question_type"
        case promptZh = "prompt_zh"
        case options
        case correctOptionKey = "correct_option_key"
        case root
        case targetQuality = "target_quality"
        case musicKey = "music_key"
        case progressionRoman = "progression_roman"
        case hintZh = "hint_zh"
        case playbackFretsSixToOne = "playback_frets_six_to_one"
    }

    /// 程序化构造（和弦听辨等），不依赖 `ear_seed` 条目。
    public init(
        id: String,
        mode: String,
        questionType: String,
        promptZh: String,
        options: [EarMcqOption],
        correctOptionKey: String,
        root: String?,
        targetQuality: String?,
        musicKey: String? = nil,
        progressionRoman: String? = nil,
        hintZh: String? = nil,
        playbackFretsSixToOne: [Int]? = nil
    ) {
        self.id = id
        self.mode = mode
        self.questionType = questionType
        self.promptZh = promptZh
        self.options = options
        self.correctOptionKey = correctOptionKey
        self.root = root
        self.targetQuality = targetQuality
        self.musicKey = musicKey
        self.progressionRoman = progressionRoman
        self.hintZh = hintZh
        self.playbackFretsSixToOne = playbackFretsSixToOne
    }
}

public struct EarSeedDocument: Decodable, Sendable {
    public let bankA: [EarBankItem]
    public let bankB: [EarBankItem]

    private struct Banks: Decodable {
        let A: [EarBankItem]?
        let B: [EarBankItem]?
    }

    private enum CodingKeys: String, CodingKey {
        case banks
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let banks = try container.decode(Banks.self, forKey: .banks)
        bankA = banks.A ?? []
        bankB = banks.B ?? []
    }
}
